tests/tests/loopback.rs

```diff
@@ -0,0 +1,306 @@
+                vni: Vni::new(1701u32).unwrap(),
+            },
+        })?;
+        Ok(())
+    }
+}
+impl Drop for Xde {
+    /// When this object is dropped, remove the xde kernel module from the
+    /// underlying system.
+    fn drop(&mut self) {
+        let mut cmd = Command::new("pfexec");
+        cmd.args(["rem_drv", "xde"]);
+        if let Err(e) = cmd.output() {
+            eprintln!("failed to remove xde driver: {}", e);
+        }
+    }
+}
+
+/// This is an xde loopback topology. There are two zones, each with a vnic
+/// sitting atop an opte device. The opte device uses a pair of simnet
+/// devices as underlay links. These simnet devices are connected to each
+/// other.
+///
+///         zone a
+///     #============#
+///     | *--------* |    *-------*
+///     | | vopte0 |------| opte0 |         *------*
+///     | *--------* |    *-------*      +--| sim0 |--+
+///     #============#       |  *-----*  |  *------*  |
+///                          +--| xde |--+            |
+///     #============#       |  *-----*  |  *------*  |
+///     | *--------* |    *-------*      +--| sim1 |--+
+///     | | vopte1 |------| opte1 |         *------*
+///     | *--------* |    *-------*
+///     #============#
+///         zone b
+///
+/// The following system of overlay/underlay routes is set up
+///
+/// 10.0.0.1 -> fd47::/74 via sim1
+/// 10.0.0.2 -> fd74::/74 via sim0
+///
+/// Zone a has an overlay address of 10.0.0.1 and zone b has an overlay
+/// address of 10.0.0.2. This means that OPTE will encap/decap packets to
+/// and from the vopte devices, the underlying routes on the host will lead
+/// xde to select the correct simnet device. Then the simnet setup will
+/// forward the packet to it's adjacent peer, which then makes it's way back
+/// to OPTE and then to the adjacent vopte device. This is a nice little
+/// sanity checker to make sure basic opte/xde functionality is working - and
+/// that we're not hitting things like debug asserts in the OS.
+#[test]
+fn test_xde_loopback() -> Result<()> {
+    // Create the "underlay loopback". With simnet device pairs, any packet that
+    // goes in one is forwarded to the other. In the topology depicted above,
+    // this means that anything vopte0 sends, will be encapsulated onto the
+    // underlay by opte0, sent to sim0 (more on the routing that makes that
+    // happen later), is forwarded to sim1, is decapsulated by opte1 and then
+    // sent to vopte1.
+    let sim = SimnetLink::new("sim0", "sim1")?;
+    let ll0 = LinkLocal::new(&sim.end_a, "ll")?;
+    let ll1 = LinkLocal::new(&sim.end_b, "ll")?;
+
+    Xde::set_xde_underlay(&sim.end_a, &sim.end_b)?;
+    // TODO this is a sort of force unset underlay until we have an unset
+    // underlay command. When this object drops it will remove the xde driver.
+    // If we do not do this, xde will hold references to the simnet devices
+    // preventing us from cleaning them up after this test.
+    let _xde = Xde {};
```

Comment:
I think this answers my question above about `rem_drv`. Is `modunload` insufficient?

Reply:
The logic here is that this test is running with a driver that was specifically added to run this test.

https://github.com/oxidecomputer/opte/blob/72bb20c8f6c740dccd072a21f09aee333eca65e1/.github/buildomat/jobs/test.sh#L34-L37

And we're doing the opposite on the way out. Admittedly, this would be better if the whole cycle was in the rust-based test harness instead of split across bash and rust.